README.md
# Light API

> A proxy API for a home automation light control system

## Usage

Clone this repository, and then install dependencies using `npm install`. You
can then start the server using `npm start`.
The server will bind to port 3000. All device data is stored in memory and only
persists as long as the server is running. If the server is restarted, any
changes will be lost.

## Documentation

All endpoints return JSON and will have one of the following top level keys:

- `data` - The response data for the request
- `errors` - An array of error objects

### Get all devices

```
GET /api/v1/device
```

**Example:**

```js
fetch('http://localhost:3000/api/v1/device')
	.then((response) => {
		console.log(response.data)
	})
```

### Get a single device

```
GET /api/v1/device/:id
```

**Example:**

```js
fetch('http://localhost:3000/api/v1/device/3')
	.then((response) => {
		console.log(response.data)
	})
```

### Update a single device

```
PATCH /api/v1/device/:id
```

**Example:**

```js
fetch('/api/v1/device/3', {
    method: 'PATCH',
		headers: {
      'content-type': 'application/json'
    },
    body: JSON.stringify({
			data: {
				brightness: 30
			}
		})
  })
  .then(response => response.json())
```



CHANGELOG.md
# Change Log

All notable changes to this project will be documented in this file
automatically by Versionist. DO NOT EDIT THIS FILE MANUALLY!
This project adheres to [Semantic Versioning](http://semver.org/).

## 0.0.2 - 2019-07-15

* Update project dependencies [Lucian]
* Shell: Fix bug on Windows shells [Lucian Buzzo]
* Tidy up headers in README [Lucian Buzzo]
* Add missing debug package [Lucian Buzzo]


test/index.spec.js
/* globals describe, it */
const chai = require('chai')
const fetch = require('node-fetch')

const {
	expect
} = chai

require('../index.js')

describe('GET /api/v1/device', () => {
	it('should return an array of devices', (done) => {
		fetch('http://localhost:3000/api/v1/device')
			.then((response) => {
				return response.json()
			})
			.then(({
				data
			}) => {
				expect(data).to.be.an('array')
				done()
			})
	})
})

describe('GET /api/v1/device:id', () => {
	it('should return a single device', (done) => {
		fetch('http://localhost:3000/api/v1/device/1')
			.then((response) => {
				return response.json()
			})
			.then(({
				data
			}) => {
				expect(data).to.be.an('object')
				expect(data).to.include({
					id: 1,
					name: 'Balcony',
					active: true,
					brightness: 50
				})
				done()
			})
	})

	it('should fail if the id is not numeric', (done) => {
		fetch('http://localhost:3000/api/v1/device/foobar')
			.then((response) => {
				expect(response.status).to.equal(400)
				done()
			})
	})

	it('should fail if the device is not found', (done) => {
		fetch('http://localhost:3000/api/v1/device/12345')
			.then((response) => {
				expect(response.status).to.equal(404)
				done()
			})
	})
})

describe('PATCH /api/v1/device:id', () => {
	it('should update a single device', (done) => {
		fetch('http://localhost:3000/api/v1/device/3', {
			method: 'PATCH',
			headers: {
				'content-type': 'application/json'
			},
			body: JSON.stringify({
				data: {
					brightness: 30
				}
			})
		})
			.then(() => {
				return fetch('http://localhost:3000/api/v1/device/3')
			})
			.then((response) => {
				return response.json()
			})
			.then(({
				data
			}) => {
				expect(data).to.be.an('object')
				expect(data).to.include({
					brightness: 30
				})
				done()
			})
	})

	it('should fail if the id is not numeric', (done) => {
		fetch('http://localhost:3000/api/v1/device/foobar', {
			method: 'PATCH',
			headers: {
				'content-type': 'application/json'
			},
			body: JSON.stringify({
				data: {
					brightness: 30
				}
			})
		})
			.then((response) => {
				expect(response.status).to.equal(400)
				done()
			})
	})

	it('should fail if the device is not found', (done) => {
		fetch('http://localhost:3000/api/v1/device/12345', {
			method: 'PATCH',
			headers: {
				'content-type': 'application/json'
			},
			body: JSON.stringify({
				data: {
					brightness: 30
				}
			})
		})
			.then((response) => {
				expect(response.status).to.equal(404)
				done()
			})
	})

	it('should fail if an update is not provided', (done) => {
		fetch('http://localhost:3000/api/v1/device/3', {
			method: 'PATCH',
			headers: {
				'content-type': 'application/json'
			}
		})
			.then((response) => {
				expect(response.status).to.equal(400)
				done()
			})
	})
})


index.js
// Set DEBUG var here for compatibility with Windows shells
process.env.DEBUG = 'express:*,light-api:*'

const bodyParser = require('body-parser')
const express = require('express')
const _ = require('lodash')
const debug = require('debug')('light-api:server')

const app = express()

app.use(bodyParser.json())

app.use((req, res, next) => {
	res.header('Access-Control-Allow-Origin', '*')
	res.header('Access-Control-Allow-Headers', [
		'Accept',
		'Authorization',
		'Content-Type',
		'Origin',
		'X-Requested-With'
	].join(', '))
	res.header('Access-Control-Allow-Methods', [
		'DELETE',
		'GET',
		'HEAD',
		'OPTIONS',
		'PATCH',
		'POST',
		'PUT'
	].join(', '))
	next()
})

const devices = [
	{
		id: 1,
		name: 'Balcony',
		active: true,
		brightness: 50
	},
	{
		id: 2,
		name: 'Bedroom 01',
		active: false,
		brightness: 70
	},
	{
		id: 3,
		name: 'Bedroom 02',
		active: false,
		brightness: 70
	},
	{
		id: 4,
		name: 'Entrance',
		active: true,
		brightness: 100
	},
	{
		id: 5,
		name: 'Kitchen',
		active: true,
		brightness: 100
	},
	{
		id: 6,
		name: 'Living Room',
		active: true,
		brightness: 40
	},
	{
		id: 7,
		name: 'Master Bedroom',
		active: false,
		brightness: 60
	},
	{
		id: 8,
		name: 'Storage',
		active: false,
		brightness: 100
	}
]

app.get('/api/v1/device', (req, res) => {
	res.json({
		data: devices
	})
})

app.get('/api/v1/device/:id', (req, res) => {
	const id = parseInt(req.params.id, 10)

	if (_.isNaN(id)) {
		return res.status(400).json({
			errors: [
				{
					detail: `Device id is not a number: ${req.params.id}`
				}
			]
		})
	}

	const device = _.find(devices, {
		id
	})

	if (!device) {
		return res.status(404).json({
			errors: [
				{
					detail: `Could not find device with id ${id}`
				}
			]
		})
	}

	return res.json({
		data: device
	})
})

app.patch('/api/v1/device/:id', (req, res) => {
	const id = parseInt(req.params.id, 10)

	if (_.isNaN(id)) {
		return res.status(400).json({
			errors: [
				{
					detail: `Device id is not a number: ${req.params.id}`
				}
			]
		})
	}

	const update = req.body

	if (!_.get(update, [ 'data' ])) {
		return res.status(400).json({
			errors: [
				{
					detail: 'No data provided for update'
				}
			]
		})
	}

	const device = _.find(devices, {
		id
	})

	if (!device) {
		return res.status(404).json({
			errors: [
				{
					detail: `Could not find device with id ${id}`
				}
			]
		})
	}

	_.merge(device, _.pick(update.data, [
		'name',
		'active',
		'brightness'
	]))

	return res.json({
		data: device
	})
})

app.listen(3000, () => {
	debug('API listening on port 3000!')
})
